### PR TITLE
Abstract storage

### PR DESCRIPTION
**Pull request by ghost**
### What does this PR do? How does it affect users?

This PR separates the specific storage we use for RabbitMQ messages that somehow failed to be delivered from the actual client. The users — that is, developers who switch to this new version — would have to also include at least one of storages as a dependency (there is only MySQL one, and it's not merged yet), and configure it separately; apart from that, no change would be necessary.

### How should this be tested (feature switches, URLs, special user permissions)?

Check that RabbitMQ messages work normally.

`/ops/core/disable` endpoint (or whatever other service uses) should stop messages from being consumed.

`/ops/core/enable` endpoint should resume consuming.

### Related Asana task, wiki page or blog posts

https://app.asana.com/0/390663436419819/470130837715893
https://github.com/gawkermedia/kinja-libs/pull/65 — MySQL-based store.
https://github.com/gawkermedia/kinja-core/pull/3159 — `core` PR to use this.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -76,138 +72,92 @@ So your options are:
...
-	protected override lazy val connection = ProductionAmqpConnection.connection
-    
-	protected override lazy val configuration: AmqpConfiguration = ProductionAmqpConfiguration
-
-	protected override lazy val messageStore: MessageStore = RabbitMQNullMessageStore
-
-	override lazy val actorSystem: ActorSystem = play.libs.Akka.system
-
-	override protected lazy val logger: Logger = LoggerFactory.getLogger(this.getClass().getName())
+class RabbitMQClientFactory extends DBComponents {
```
Comment (balagez) [inline README.md]: You should make all documentation and code not make any assumptions about the environment the library is being used and what message store is implemented. So this example should not contain any Play Framework specific snippets, nor a MySQL implementation. Make everything abstract, use pseudo code if necessary.

File: build.sbt [hunk 2]
```diff
@@ -4,7 +4,7 @@ import org.scalastyle.sbt.ScalastylePlugin
 
 name := "amqp-client-provider"
 
-version := "2.3.0"
+version := "2.4.0"
```
Comment (balagez) [inline build.sbt]: This is a major, incompatible change. Version should be `3.0.0`.

File: src/main/scala/com/kinja/amqp/persistence/InMemoryMessageBufferDecorator.scala [hunk 3]
```diff
@@ -33,34 +33,32 @@ class InMemoryMessageBufferDecorator(
...
-			inMemoryMessageBuffer ! MultipleConfirmation(confirm)
-		} else {
-			messageStore.saveConfirmation(confirm)
-		}
+	override def saveConfirmations(confirms: List[MessageConfirmation]): Unit = {
+		val (multiples, singles) = confirms.partition(_.multiple)
+		// we don't save every multiple confirmation here,
+		// just collect (and increment) them and save all at once in the flush loop
+		inMemoryMessageBuffer ! MultipleConfirmations(multiples)
+		messageStore.saveConfirmations(singles)
```
Comment (balagez) [inline src/main/scala/com/kinja/amqp/persistence/InMemoryMessageBufferDecorator.scala]: You should probably
* only send `MultipleConfirmations` if `multiples.nonEmpty`
* only call `saveConfirmations` if `singles.nonEmpty`

File: src/main/scala/com/kinja/amqp/persistence/InMemoryMessageBufferDecorator.scala [hunk 4]
```diff
@@ -33,34 +33,32 @@ class InMemoryMessageBufferDecorator(
...
-	override def lockRowsOlderThan(olderThanSeconds: Long, lockTimeOutAfterSeconds: Long, limit: Int): Int = {
-		messageStore.lockRowsOlderThan(olderThanSeconds, lockTimeOutAfterSeconds, limit)
+	override def lockOldRows(limit: Int): Int = {
+		messageStore.lockOldRows(limit)
 	}
 
-	override def saveMessage(msg: Message): Unit = {
-		inMemoryMessageBuffer ! SaveMessage(msg)
+	override def saveMessages(msgs: List[Message]): Unit = {
+		inMemoryMessageBuffer ! SaveMessages(msgs)
```
Comment (balagez) [inline src/main/scala/com/kinja/amqp/persistence/InMemoryMessageBufferDecorator.scala]: Only do this if `msgs.nonEmpty`?

File: src/main/scala/com/kinja/amqp/AmqpConfiguration.scala [hunk 5]
```diff
@@ -14,23 +14,13 @@ final case class ResendLoopConfig(
...
 )
 
-sealed abstract class DeliveryGuarantee(val configValue: String) extends Product with Serializable
-object DeliveryGuarantee {
-	case object AtLeastOnce extends DeliveryGuarantee("at-least-once")
-	case object AtMostOnce extends DeliveryGuarantee("at-most-once")
-}
-
-final case class ProducerConfig(deliveryGuarantee: DeliveryGuarantee, exchangeParams: ExchangeParameters)
+final case class ProducerConfig(deliveryGuarantee: String, exchangeParams: ExchangeParameters)
```
Comment (balagez) [inline src/main/scala/com/kinja/amqp/AmqpConfiguration.scala]: Why do you need to change the type safe enum to a String that might have so many problems? Please keep the enum. I also dislike the concept of empty string as default value. As a user of this library, what does it tell you? What delivery guarantee the library will provide? What's "built-in"?

File: src/main/scala/com/kinja/amqp/AmqpClientFactory.scala [hunk 6]
```diff
@@ -15,25 +15,23 @@ class AmqpClientFactory {
 		actorSystem: ActorSystem,
 		logger: Logger,
 		ec: ExecutionContext,
-		getWriteConnection: () => Connection,
-		getReadConnecion: () => Connection,
-		hostname: String
+		messageStores: Map[String, MessageStore]
```
Comment (balagez) [inline src/main/scala/com/kinja/amqp/AmqpClientFactory.scala]: What's the String here? I know what it is, but as a user of the API who's unfamiliar with it, a String parameter doesn't tell anything. Is it the message store name maybe?

File: src/main/scala/com/kinja/amqp/AmqpClientFactory.scala [hunk 7]
```diff
@@ -15,25 +15,23 @@ class AmqpClientFactory {
...
 	): AmqpClientInterface =
 		{
 			if (config.testMode) {
 				new NullAmqpClient
 			} else {
 				val connection: ActorRef = createConnection(config, actorSystem)
-				val messageStore: MessageStore = createMessageStore(
-					config, actorSystem, logger, ec, getWriteConnection, getReadConnecion, hostname
+				val bufferedMessageStores = messageStores.mapValues(createMessageStore(
```
Comment (balagez) [inline src/main/scala/com/kinja/amqp/AmqpClientFactory.scala]: Never use `mapValues`!!! That should be removed from the Scala library. It only creates a view of the original map with the function applied to items every single f*ing time they're accessed. So here `createMessageStore` will be called each and every time you access a value from `bufferedMessageStores`.

File: src/main/scala/com/kinja/amqp/AmqpClient.scala [hunk 8]
```diff
@@ -13,16 +13,37 @@ class AmqpClient(
 	val actorSystem: ActorSystem,
 	private val configuration: AmqpConfiguration,
 	private val logger: Slf4jLogger,
-	private val messageStore: MessageStore,
+	private val messageStores: Map[String, MessageStore],
```
Comment (balagez) [inline src/main/scala/com/kinja/amqp/AmqpClient.scala]: As a user of this library I'd ask what's the `String` here? It should be an enum at least. Why is this a `Map` and not a `List[(DeliveryGuarantee, MessageStore)]`, so essentially why can you only have one message store for a given type of guarantee? Please make such choices better explained by using better designed types. `Map` and `String` are just too generic.

**Review by balagez — CHANGES_REQUESTED** [review]
I didn't check every single line but left enough comments until the next round of review.

File: src/main/scala/com/kinja/amqp/AmqpClientFactory.scala (hunk 7 above)
Comment (ghost) [reply src/main/scala/com/kinja/amqp/AmqpClientFactory.scala]: Oh, shit, I wasn't aware of that. It's a landmine indeed. Thanks!

File: src/main/scala/com/kinja/amqp/AmqpConfiguration.scala (hunk 5 above)
Comment (ghost) [reply src/main/scala/com/kinja/amqp/AmqpConfiguration.scala]: This is now extensible, so `enum` won't do. But we can use `case class DeliveryGuarantee(whatever: String)`, if you think it's better.

File: README.md (hunk 1 above)
Comment (ghost) [reply README.md]: Well, I think we should at least make a reference to some existing implementation — MySQL probably, as Redis is not yet complete.

File: build.sbt (hunk 2 above)
Comment (ghost) [reply build.sbt]: OK. Will do the fixes tomorrow.

File: src/main/scala/com/kinja/amqp/AmqpConfiguration.scala (hunk 5 above)
Comment (ghost) [reply src/main/scala/com/kinja/amqp/AmqpConfiguration.scala]: I'm not sure about empty string stuff myself, but we shouldn't require massive changes in configuration for most projects. Which means that some reasonable default should be used.

I can probably make `DeliveryGuarantee` to be an `Option[String]`.

Comment (ghost) [reply src/main/scala/com/kinja/amqp/AmqpConfiguration.scala]: As for built-in, I really don't know how to make it more understandable. We need something for built-in exchanges, like `amq.topic`. That's it.

File: src/main/scala/com/kinja/amqp/persistence/InMemoryMessageBufferDecorator.scala (hunk 4 above)
Comment (ghost) [reply src/main/scala/com/kinja/amqp/persistence/InMemoryMessageBufferDecorator.scala]: Done.

File: src/main/scala/com/kinja/amqp/persistence/InMemoryMessageBufferDecorator.scala (hunk 3 above)
Comment (ghost) [reply src/main/scala/com/kinja/amqp/persistence/InMemoryMessageBufferDecorator.scala]: Done.

File: src/main/scala/com/kinja/amqp/AmqpClientFactory.scala (hunk 7 above)
Comment (ghost) [reply src/main/scala/com/kinja/amqp/AmqpClientFactory.scala]: Done. Removed `mapValues` in several other places, just to be on the safe side.

Is there a wart for that?

File: src/main/scala/com/kinja/amqp/AmqpClientFactory.scala (hunk 6 above)
Comment (ghost) [reply src/main/scala/com/kinja/amqp/AmqpClientFactory.scala]: Made it into `DeliveryGuarantee`.

File: src/main/scala/com/kinja/amqp/AmqpClient.scala (hunk 8 above)
Comment (ghost) [reply src/main/scala/com/kinja/amqp/AmqpClient.scala]: I understand your objection against `String`, but `Map`? One message store for one type of guarantee is exactly what we need.

File: build.sbt (hunk 2 above)
Comment (ghost) [reply build.sbt]: Done.

File: README.md (hunk 1 above)
Comment (ghost) [reply README.md]: Fiddled with it a bit.

File: src/main/scala/com/kinja/amqp/AmqpConfiguration.scala (hunk 5 above)
Comment (ghost) [reply src/main/scala/com/kinja/amqp/AmqpConfiguration.scala]: The library does not provide delivery guarantees. That's the point. Message stores do, the library just enables them to do so.

File: src/main/scala/com/kinja/amqp/AmqpClient.scala (hunk 8 above)
Comment (balagez) [reply src/main/scala/com/kinja/amqp/AmqpClient.scala]: This was exactly the kind of decision that was not obvious for me. Maybe some documentation could explain this.

Comment (ghost) [reply src/main/scala/com/kinja/amqp/AmqpClient.scala]: I've added a few words about that to the README.

File: README.md [hunk 9]
```diff
@@ -76,138 +72,88 @@ So your options are:
...
 * `exchanges`: The list of exchanges you would like to use. Built in exchanges (amq.direct, amq.topic, etc.) are included by default, you don't have to add them here. The index of the exchange config will be the name of the exchange. With every exchange, you can configure:
   * `type`: The type of exchange (direct, topic, fanout, headers) 
-  * `deadLetterExchange`: The name of dead letter exchange for the exchange. You have to configure that here also, or you can use on of the built in exchanges. 
+  * `deadLetterExchange`: The name of dead letter exchange for the exchange. You have to configure that here also, or you can use on of the built in exchanges.
+  * `atLeastOnce`: Delivery guarantee used for this exchange. Default value is assumed if this is omitted. It's referred in code as `AtLeastOnceGroup.default`.
+* `builtinAtLeastOnce`: Delivery guarantee used for several predefined exchanges, like `amq.topic`. Default value is assumed if this is omitted. It's referred in code as `AtLeastOnceGroup.default`.
```
Comment (balagez) [inline README.md]: Documentation here should probably detail what this "group" concept means. That you can implement various message stores, and you can group exchanges into delivery guarantee groups based on what delivery guarantee the message store provides. Or something like this. Examples would be useful too. And shouldn't we call these config keys `atLeastOnceGroup` and `builtinAtLeastOnceGroup` respectively? Make sure to also rename occurrence in `AmqpConfiguration.scala`.

File: src/main/scala/com/kinja/amqp/AmqpClient.scala [hunk 10]
```diff
@@ -13,16 +13,37 @@ class AmqpClient(
...
-	private var repeater: Option[MessageBufferProcessor] = None
+	private lazy val repeater: List[MessageBufferProcessor] =
+		messageStores.toList.map {
+			case (guarantee, messageStore) =>
+				val conf = configuration.resendConfig.getOrElse(throw new MissingResendConfigException)
+				val selectedProducers = producers.filterKeys { exchangeName =>
+					configuration.exchanges
+						.get(exchangeName)
+						.map(_.atLeastOnceGroup)
+						.contains(guarantee)
```
Comment (balagez) [inline src/main/scala/com/kinja/amqp/AmqpClient.scala]: Maybe rename `guarantee` to `group` for clarity?

**Review by balagez — APPROVED** [review]
Minor comments left, otherwise looks good.

**Comment by ghost** [comment]
It's a library, so, it won't go out unless someone actually uses it. Merging.

**Comment by balagez** [comment]
Will you please create a release? https://github.com/gawkermedia/amqp-client-provider/releases

**Comment by ghost** [comment]
Right. Forgot about it. Thanks.

**Pull request merged**